### PR TITLE
Authorization support for OIDC

### PR DESCRIPTION
**Pull request by universam1**
# AuthZ filter OIDCQueryClaimsFilter

This filter is intended to be chained after one of `OAuthOidc*Claims` OIDC flow provider, making use of the token information saved in `stateBag` to classify **authorization**

* It leverages the query syntax of `GJSON` to express a matching of the token.
* It allows validation per path request, matching one or more queries.
* Each argument of the filter expresses a path/query relation, one or several paths can be defined
* The first positive match wins

Format:
`<path>:[<query>]`

Following example statements would be possible:
```
"/some/path:@_:sub==\"somesub\")"
"/some/path:[@this].#(sub==\"somesub\")"
"/another/path:groups.#(%\"CD-*\")"
"/asdf/:groups.#(%\"CD-*\") 'groups.#(%\"*-Test-Users\")' groups.#(%\"Purchasing-Department\")"
"/:groups.#(%\"Purchasing-Department\")"
```

These would match each one of these:
```json
{
    "sub": "somesub",
    "groups": [
        "CD-Administrators",
        "Purchasing-Department",
        "AppX-Test-Users",
    ]
}
```
ref: https://github.com/tidwall/gjson/issues/151

## generalize and remove cognitive challenge

 * Cognitive challenge reduced by splitting out the stage 5 of the OIDC flow into distinct method
 * generlizing duplicate code paths
 * making use of one struct `tokenContainer` as they can be united through `omitempty` annotation. This is necessary for later filter chain parsing of the statebag

### CODE REVIEW

**Comment by universam1** [comment]
Again, this is not yet ready @aryszka @szuecs , however added a benchmark and the results are pretty much equal to existing plain validations, while a magnitude more flexible

```
goos: darwin
goarch: amd64
pkg: github.com/zalando/skipper/filters/auth
Benchmark_OIDCSetup/Query_group_claim,_all_claims_valid-16         	      92	  12578983 ns/op	 1598629 B/op	    2935 allocs/op
Benchmark_OIDCSetup/UserInfo_valid_claims_requested-16             	      90	  12676063 ns/op	 1611687 B/op	    2985 allocs/op
Benchmark_OIDCSetup/AllClaims_all_claims_valid-16                  	      93	  12265713 ns/op	 1583038 B/op	    2895 allocs/op
Benchmark_OIDCSetup/AnyClaims,_any_valid-16                        	      96	  12184460 ns/op	 1573918 B/op	    2896 allocs/op
PASS
ok  	github.com/zalando/skipper/filters/auth	142.770s
Success: Benchmarks passed.
```

**Comment by dhohengassner** [comment]
looks like this would be a huge feature improvement 😮 
👍 would love to see this released

File: filters/auth/oidc_introspection.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	return fmt.Sprintf("%s(%s)", OidcClaimsQueryName, strings.Join(str, "; "))
+}
+
+func (filter *oidcIntrospectionFilter) Request(ctx filters.FilterContext) {
+	r := ctx.Request()
+
+	token, ok := ctx.StateBag()[oidcClaimsCacheKey].(tokenContainer)
+	if !ok || &token == (&tokenContainer{}) || len(token.Claims) == 0 {
+		log.Errorf("Error retrieving oidcClaimsCacheKey for OIDC token introspection")
+		unauthorized(ctx, "", missingToken, r.Host, "oidcClaimsCacheKey unavailable")
```
Comment (szuecs) [inline filters/auth/oidc_introspection.go]: oidcClaimsCacheKey as string should use the defined const (also line 101)

File: filters/auth/oidc_introspection_test.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+			got, err := spec.CreateFilter(tt.args)
+			tt.wantErr(t, err)
+			assert.Equal(t, tt.want, got)
+		})
+	}
+
+}
+
+func TestOIDCQueryClaimsFilter(t *testing.T) {
+	// logrus.SetLevel(logrus.DebugLevel)
```
Comment (szuecs) [inline filters/auth/oidc_introspection_test.go]: please drop the comment :)

File: filters/auth/oidc_introspection_test.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+			defer oidcServer.Close()
+			t.Logf("oidc/auth server URL: %s", oidcServer.URL)
+			// create filter
+			sargs := []interface{}{
+				validClient,
+				"mysec",
+				proxy.URL + "/redirect",
+			}
+			sargs = append([]interface{}{oidcServer.URL}, sargs...)
+			sargs = append(sargs, testKey, testKey)
```
Comment (szuecs) [inline filters/auth/oidc_introspection_test.go]: Any reason why not:
```
sargs := []interface{}{
	oidcServer.URL,  
	validClient,
	"mysec",
	proxy.URL + "/redirect",
	testKey,
	testKey,
}
```

File: go.mod [hunk 4]
```diff
@@ -15,6 +15,7 @@ require (
 	github.com/go-redis/redis/v7 v7.0.0-beta.4
 	github.com/go-yaml/yaml v2.1.0+incompatible
 	github.com/google/go-cmp v0.3.0
+	github.com/gotestyourself/gotestyourself v2.2.0+incompatible // indirect
```
Comment (szuecs) [inline go.mod]: I did not found where the dependency is coming from, can you please check if `go mod tidy` would drop it?

File: go.mod [hunk 5]
```diff
@@ -44,6 +47,7 @@ require (
 	google.golang.org/grpc v1.22.0 // indirect
 	gopkg.in/square/go-jose.v2 v2.3.1
 	gopkg.in/yaml.v2 v2.2.3
+	gotest.tools v2.2.0+incompatible // indirect
```
Comment (szuecs) [inline go.mod]: I did not found where the dependency is coming from, can you please check if `go mod tidy` would drop it?

**Comment by szuecs** [comment]
@universam1 Code looks great to me, some smaller comments to fix and please check the dependencies, no idea where this came from.

File: filters/auth/oidc_introspection_test.go (hunk 3 above)
Comment (universam1) [reply filters/auth/oidc_introspection_test.go]: That's a very valid point - guess copy-pasta 😄 thanks

File: filters/auth/oidc_introspection_test.go (hunk 2 above)
Comment (universam1) [reply filters/auth/oidc_introspection_test.go]: 👍

File: filters/auth/oidc_introspection.go (hunk 1 above)
Comment (universam1) [reply filters/auth/oidc_introspection.go]: makes sense!

File: go.mod (hunk 4 above)
Comment (universam1) [reply go.mod]: Thanks for double checking - littered from coding. Cleaned it up

File: go.mod (hunk 5 above)
Comment (universam1) [reply go.mod]: thanks for the heads up - yes `go mod tidy` cleaned it 👍

**Comment by universam1** [comment]
> some smaller comments to fix and please check the dependencies, no idea where this came from.

All remarks implemented @szuecs , thank you for the review! Hope it's ready for merge

**Comment by szuecs** [comment]
:+1:

File: docs/tutorials/auth.md [hunk 6]
```diff
@@ -223,3 +223,25 @@ for which it needs a secret key. This key is in a file which can be rotated peri
 because it is reread by Skipper. The path to this file can be passed with the flag
 `-oidc-secret-file` when Skipper is started.
 
+### Authorization and access control
+
+Authorization validation and access control is available by means of a subsequent filter `oidcClaimsQuery`. It inspects the ID token, that is present after a successful `oauthOidc*` filter step, and validates the defined query with the request path.
```
Comment (icereed) [inline docs/tutorials/auth.md]: Hi there @universam1,

should we include a link to the reference here? Might be easier to deep-dive then.

Comment (universam1) [reply docs/tutorials/auth.md]: sure, added a link

**Comment by universam1** [comment]
> 👍

added documentation, references and examples. Ready merge @szuecs

**Comment by aryszka** [comment]
:+1:

**Comment by szuecs** [comment]
:+1:

**Comment by szuecs** [comment]
Thanks, great step forward for authz in skipper!

**Comment by universam1** [comment]
Thank you @szuecs  for merging - just by chance is there a T-shirt bounty program or so available from Zalando? ;-)

**Comment by szuecs** [comment]
@universam1 I tried to figure out if we have a contributor bounty program, but I did not get a positive answer, yet. :(

**Pull request merged**